internal/overlord/cmdstate/cmdmgr.go

```diff
@@ -0,0 +1,654 @@
+		cmd.SysProcAttr.Setctty = true
+	}
+
+	finisher := func(exitCode int, cmdErr error) error {
+		for _, tty := range ttys {
+			_ = tty.Close()
+		}
+
+		// Close the control channel, if connected.
+		controlConn := e.getWebsocket(wsControl)
+		if controlConn != nil {
+			_ = controlConn.Close()
+		}
+
+		close(childDead)
+
+		wg.Wait()
+
+		for _, pty := range ptys {
+			_ = pty.Close()
+		}
+
+		setApiData(change, exitCode)
+
+		return cmdErr
+	}
+
+	// Start the command!
+	err = cmd.Start()
+	if err != nil {
+		return finisher(-1, err)
+	}
+
+	// Send its PID to the control loop.
+	pidCh <- cmd.Process.Pid
+
+	// Wait for it to finish.
+	err = cmd.Wait()
+
+	// Handle errors: timeout, non-zero exit code, or other error.
+	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+		return finisher(-1, fmt.Errorf("timed out after %v: %w", e.timeout, ctx.Err()))
+	} else if exitErr, ok := err.(*exec.ExitError); ok {
+		status, ok := exitErr.Sys().(syscall.WaitStatus)
+		if ok {
+			if status.Signaled() {
+				// 128 + n == Fatal error signal "n"
+				return finisher(128+int(status.Signal()), nil)
+			}
+			return finisher(status.ExitStatus(), nil)
+		}
+		return finisher(-1, err)
+	} else if err != nil {
+		return finisher(-1, err)
+	}
+
+	// Successful exit (exit code 0).
+	return finisher(0, nil)
+}
+
+func setApiData(change *state.Change, exitCode int) {
+	st := change.State()
+	st.Lock()
+	defer st.Unlock()
+	change.Set("api-data", map[string]interface{}{
+		"exit-code": exitCode,
+	})
+}
+
+type execCommand struct {
+	Command string          `json:"command"`
+	Signal  *execSignalArgs `json:"signal"`
+	Resize  *execResizeArgs `json:"resize"`
+}
+
+type execSignalArgs struct {
+	Name string `json:"name"`
+}
+
+type execResizeArgs struct {
+	Width  int `json:"width"`
+	Height int `json:"height"`
+}
+
+func (e *execution) controlLoop(pidCh <-chan int, exitCh <-chan struct{}, ptyFd int) {
+	logger.Debugf("Control handler waiting")
+	defer logger.Debugf("Control handler finished")
+
+	// Wait till we receive the process's PID (command started).
+	var pid int
+	select {
+	case pid = <-pidCh:
+		break
+	case <-exitCh:
+		return
+	}
+
+	// Wait till the control websocket is connected.
+	select {
+	case <-e.controlConnected:
+		break
+	case <-exitCh:
+		return
+	}
+
+	logger.Debugf("Control handler started for child PID %d", pid)
+	for {
+		controlConn := e.getWebsocket(wsControl)
+		mt, r, err := controlConn.NextReader()
+		if mt == websocket.CloseMessage {
+			break
+		}
+
+		if err != nil {
+			logger.Debugf("Cannot get next reader for PID %d: %v", pid, err)
+			er, ok := err.(*websocket.CloseError)
+			if !ok {
+				break
+			}
+			if er.Code != websocket.CloseAbnormalClosure {
+				break
+			}
+
+			// If an abnormal closure occurred, kill the attached process.
+			err := unix.Kill(pid, unix.SIGKILL)
+			if err != nil {
+				logger.Noticef("Cannot send SIGKILL to pid %d", pid)
+			} else {
+				logger.Noticef("Sent SIGKILL to pid %d", pid)
```

Comment:
Is this worth noticing?  Looks like Debug.

Reply:
Done.